dali/operators/coord/coord_flip.cu

```diff
@@ -0,0 +1,129 @@
+  int64_t tid = threadIdx.y * blockDim.x + threadIdx.x;
+  for (int64_t idx = offset + tid; idx < sample.size; idx += grid_size) {
+    int d = idx % ndim;
+    bool flip = static_cast<bool>(sample.flip_dim_mask & (1 << d));
+    sample.out[idx] = flip ? T(1) - sample.in[idx] : sample.in[idx];
+  }
+}
+
+}  // namespace
+
+class CoordFlipGPU : public CoordFlip<GPUBackend> {
+ public:
+  explicit CoordFlipGPU(const OpSpec &spec)
+      : CoordFlip<GPUBackend>(spec) {}
+
+  ~CoordFlipGPU() override = default;
+  DISABLE_COPY_MOVE_ASSIGN(CoordFlipGPU);
+
+  void RunImpl(workspace_t<GPUBackend> &ws) override;
+
+  USE_OPERATOR_MEMBERS();
+  using Operator<GPUBackend>::RunImpl;
+  using CoordFlip<GPUBackend>::layout_;
+
+ private:
+  std::vector<SampleDesc<float>> sample_descs_;
+  Tensor<GPUBackend> scratchpad_;
+};
+
+void CoordFlipGPU::RunImpl(workspace_t<GPUBackend> &ws) {
+  const auto &input = ws.InputRef<GPUBackend>(0);
+  DALI_ENFORCE(input.type().id() == DALI_FLOAT, "Input is expected to be float");
+
+  auto &output = ws.OutputRef<GPUBackend>(0);
+
+  if (layout_.empty()) {
+    layout_ = ndim_ == 2 ? "xy" : "xyz";
+  }
+
+  int x_dim = layout_.find('x');
+  DALI_ENFORCE(x_dim >= 0, "Dimension \"x\" not found in the layout");
+
+  int y_dim = layout_.find('y');
+  if (ndim_ > 1)
+    DALI_ENFORCE(y_dim >= 0, "Dimension \"y\" not found in the layout");
+
+  int z_dim = layout_.find('z');
+  if (ndim_ > 2)
+    DALI_ENFORCE(z_dim >= 0, "Dimension \"z\" not found in the layout");
+
+  sample_descs_.clear();
+  sample_descs_.reserve(batch_size_);
+  for (int sample_id = 0; sample_id < batch_size_; sample_id++) {
+    SampleDesc<float> sample_desc;
+    sample_desc.in = input.tensor<float>(sample_id);
+    sample_desc.out = output.mutable_tensor<float>(sample_id);
+    sample_desc.size = volume(input.tensor_shape(sample_id));
+    assert(sample_desc.size == volume(output.tensor_shape(sample_id)));
+
+    bool horizontal_flip = spec_.GetArgument<int>("horizontal", &ws, sample_id);
+    bool vertical_flip = spec_.GetArgument<int>("vertical", &ws, sample_id);
+    bool depthwise_flip = spec_.GetArgument<int>("depthwise", &ws, sample_id);
+
+    if (horizontal_flip) {
+      sample_desc.flip_dim_mask |= (1 << x_dim);
+    }
+
+    if (vertical_flip) {
+      sample_desc.flip_dim_mask |= (1 << y_dim);
+    }
+
+    if (depthwise_flip) {
+      sample_desc.flip_dim_mask |= (1 << z_dim);
+    }
+    sample_descs_.emplace_back(std::move(sample_desc));
+  }
+
+  scratchpad_.set_type(TypeInfo::Create<uint8_t>());
+  int64_t sz = batch_size_ * sizeof(SampleDesc<float>);
+  scratchpad_.Resize({sz});
+  auto sample_descs_gpu_ = reinterpret_cast<SampleDesc<float>*>(scratchpad_.mutable_data<uint8_t>());
+  auto stream = ws.stream();
+  CUDA_CALL(
+    cudaMemcpyAsync(sample_descs_gpu_, sample_descs_.data(), sz, cudaMemcpyHostToDevice, stream));
+
+  dim3 block(32, 32);
+  auto blocks_per_sample = std::max(32, 1024 / batch_size_);
```

Comment:
What's the reason for this gridDim.x?

Reply:
it's just a good enough number. The second term is meant to reduce the number of total blocks in case of a batch with many samples